File: app/src/main/java/org/wikipedia/json/CookieManagerTypeAdapter.java
package org.wikipedia.json;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import org.wikipedia.dataclient.SharedPreferenceCookieManager;
import org.wikipedia.dataclient.WikiSite;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.Cookie;
import okhttp3.HttpUrl;

public class CookieManagerTypeAdapter extends TypeAdapter<SharedPreferenceCookieManager> {
    @Override public void write(JsonWriter out, SharedPreferenceCookieManager cookies) throws IOException {
        Map<String, List<Cookie>> map = cookies.getCookieJar();
        out.beginObject();
        for (String key : map.keySet()) {
            out.name(key).beginArray();
            for (Cookie cookie : map.get(key)) {
                out.value(cookie.toString());
            }
            out.endArray();
        }
        out.endObject();
    }

    @Override public SharedPreferenceCookieManager read(JsonReader in) throws IOException {
        Map<String, List<Cookie>> map = new HashMap<>();
        in.beginObject();
        while (in.hasNext()) {
            String key = in.nextName();
            List<Cookie> list = new ArrayList<>();
            map.put(key, list);
            in.beginArray();
            HttpUrl url = HttpUrl.parse(WikiSite.DEFAULT_SCHEME + "://" + key);
            while (in.hasNext()) {
                String str = in.nextString();
                if (url != null) {
                    list.add(Cookie.parse(url, str));
                }
            }
            in.endArray();
        }
        in.endObject();
        return new SharedPreferenceCookieManager(map);
    }
}

File: app/src/main/java/org/wikipedia/json/CookieManagerTypeAdapter.kt
package org.wikipedia.json

import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import okhttp3.Cookie
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import org.wikipedia.dataclient.SharedPreferenceCookieManager
import org.wikipedia.dataclient.WikiSite
import java.io.IOException

class CookieManagerTypeAdapter : TypeAdapter<SharedPreferenceCookieManager>() {

    @Throws(IOException::class)
    override fun write(writer: JsonWriter, cookies: SharedPreferenceCookieManager) {
        val map = cookies.cookieJar
        writer.beginObject()
        for (key in map.keys) {
            writer.name(key).beginArray()
            map[key]?.forEach { cookie ->
                writer.value(cookie.toString())
            }
            writer.endArray()
        }
        writer.endObject()
    }

    @Throws(IOException::class)
    override fun read(reader: JsonReader): SharedPreferenceCookieManager {
        val map = mutableMapOf<String, MutableList<Cookie>>()
        reader.beginObject()
        while (reader.hasNext()) {
            val key = reader.nextName()
            val list = mutableListOf<Cookie>()
            map[key] = list
            reader.beginArray()
            val url = (WikiSite.DEFAULT_SCHEME + "://" + key).toHttpUrlOrNull()
            while (reader.hasNext()) {
                url?.let {
                    Cookie.parse(it, reader.nextString())?.run { list.add(this) }
                }
            }
            reader.endArray()
        }
        reader.endObject()
        return SharedPreferenceCookieManager(map)
    }
}
